### PR TITLE
Add Optimal Path to Score Feature

### PR DESCRIPTION
**Pull request by brandonabela**

### CODE REVIEW

File: Logic/dijkstra_graph.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+        return []
+
+    def find_a_goal_path(self, initial, end):
+        '''Responsible for calculating the optimal path for a goal'''
+
+        shortest_paths = {initial: (None, 0)}
+        current_node = initial
+        visited = set()
+
```
Comment (KeithGeorgeCiantar) [inline Logic/dijkstra_graph.py:55]: Maybe add a small comment that Dijkstra is used to get the nodes in the shortest path

File: Logic/ball_path.py [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+        l_c = (point_two[1] * point_one[0]) - (point_one[1] * point_two[0])
+
+        return (l_a, l_b, l_c)
+
+    @staticmethod
+    def line_intercept_circle(line_terms, circle_point, circle_radius):
+        '''Checks if a line intercepts a circle'''
+
+        l_a = line_terms[0]
+        l_b = line_terms[1]
```
Comment (KeithGeorgeCiantar) [inline Logic/ball_path.py]: Do a and b refer to x and y or are they like 'ax + by + c = 0'?

File: Logic/ball_path.py [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+        l_c = line_terms[2]
+
+        c_x = circle_point[0]
+        c_y = circle_point[1]
+
+        distance = abs(l_a * c_x + l_b * c_y + l_c) / math.sqrt(l_a * l_a + l_b * l_b)
+
+        return distance >= circle_radius
```
Comment (KeithGeorgeCiantar) [inline Logic/ball_path.py]: Does this take into consideration cases like the attached?

Example

Also should it return true if it intercepts or false if it intercepts? Currently it returns false if it intercepts.

File: Logic/ball_path.py [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+    ''' Responsible for finding an optimal path '''
+
+    min_angle = 40
+    max_angle = 300
+
+    def find_path(self, balls, holes, ball_colour, hit_number):
+        '''Responsible for calculating an optimal path for a number of hits'''
+
+        white_index = self.get_balls_index(balls, BallColour.White)
+        target_indexes = self.get_balls_index(balls, ball_colour)
```
Comment (KeithGeorgeCiantar) [inline Logic/ball_path.py]: Indices or Indexes? I am not sure and I have found this https://stackoverflow.com/questions/1378781/proper-terminology-should-i-say-indexes-or-indices which has some conflicting arguments. I think it's mainly important to choose one standard and use it throughout.

File: Logic/ball_path.py [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+                else:
+                    angle_weight = (angle - 180) * 1000
+
+            distance = self.distance_from_two_points(all_objects[start_index], all_objects[end_index])
+            graph.add_edge(start_index, end_index, angle_weight + distance)
+
+        return is_valid
+
+    def is_path_ocluded(self, start_position, end_position, path_indexes, all_objects):
+        '''Determine if path is ocluded by other balls'''
```
Comment (KeithGeorgeCiantar) [inline Logic/ball_path.py]: Should be 'occluded' not 'ocluded'.

File: Logic/ball_path.py [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+
+        line = self.line_equation_from_two_points(start_position, end_position)
+
+        for ball_index, a_ball in enumerate(all_objects):
+            if (ball_index not in path_indexes and len(a_ball) == 3):
+                if not self.line_intercept_circle(line, a_ball, lookup.BALL_DIAMETER):
```
Comment (KeithGeorgeCiantar) [inline Logic/ball_path.py]: If the method takes radius, why are you using diameter?

**Review by KeithGeorgeCiantar — CHANGES_REQUESTED** [review]
Some minor changes requested, otherwise everything seems to be ok.

File: Logic/dijkstra_graph.py (hunk 1 above)
Comment (brandonabela) [reply Logic/dijkstra_graph.py:55]: Done

File: Logic/ball_path.py (hunk 2 above)
Comment (brandonabela) [reply Logic/ball_path.py]: Done

File: Logic/ball_path.py (hunk 3 above)
Comment (brandonabela) [reply Logic/ball_path.py]: Done

File: Logic/ball_path.py (hunk 4 above)
Comment (brandonabela) [reply Logic/ball_path.py]: Done

File: Logic/ball_path.py (hunk 5 above)
Comment (brandonabela) [reply Logic/ball_path.py]: Done

File: Logic/ball_path.py (hunk 6 above)
Comment (brandonabela) [reply Logic/ball_path.py]: Done

**Review by KeithGeorgeCiantar — APPROVED** [review]
Worked together on the new version. The new changes seems to work well.

**Pull request merged**